utils/zerovec/src/map/vecs.rs

```diff
@@ -0,0 +1,126 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::ule::*;
+use crate::VarZeroVec;
+use crate::ZeroVec;
+use std::mem;
+
+/// Trait abstracting over [`ZeroVec`] and [`VarZeroVec`], for use in [`ZeroMap`]. You
+/// should not be implementing or calling this trait directly.
+pub trait ZeroVecLike<'a, T> {
+    /// The type received by `Self::binary_search()`
+    type NeedleType: ?Sized;
+    /// The type returned by `Self::get()`
+    type GetType: ?Sized;
+    /// Search for a key in a sorted vector, returns `Ok(index)` if found,
+    /// returns `Err(insert_index)` if not found, where `insert_index` is the
+    /// index where it should be inserted to maintain sort order.
+    fn binary_search(&self, k: &Self::NeedleType) -> Result<usize, usize>;
+    /// Get element at `index`
+    fn get(&self, index: usize) -> Option<&Self::GetType>;
+    /// Insert an element at `index`
+    fn insert(&mut self, index: usize, value: T);
+    /// Remove the element at `index` (panicking if nonexistant)
+    fn remove(&mut self, index: usize) -> T;
+    /// Replace the element at `index` with another one, returning the old element
+    fn replace(&mut self, index: usize, value: T) -> T;
```

Comment:
Returning an `impl Trait` in a trait is really hard right now, unfortunately, so we'd have to do some manual `Iterator` impls so it ends up being much more complicated for functions that get called exactly once :smile:

Reply:
> Returning an `impl Trait` in a trait is really hard right now, unfortunately, so we'd have to do some manual `Iterator` impls so it ends up being much more complicated for functions that get called exactly once 😄

Wrong thread?

Reply:
Ping

Reply:
No, correct thread, `splice()` returns an iterator, and we'd have to either return `impl Iterator` (not possible with traits) or return `struct VarZeroVecSpliceIterator` which is a lot of additional code

Reply:
What I was actually thinking was a poor-man's splice without iterators.

But, I'll retract this suggestion, because I realized that these methods are specifically the ones you need to implement the ZeroMap mutation functions.  So it's not useful to add `splice()`.